core/src/services/fs/lister.rs

```diff
@@ -93,17 +103,61 @@ impl oio::BlockingList for FsLister<std::fs::ReadDir> {
         // (no extra system calls needed), but some Unix platforms may
         // require the equivalent call to symlink_metadata to learn about
         // the target file type.
-        let file_type = de.file_type().map_err(new_std_io_error)?;
 
-        let entry = if file_type.is_file() {
-            oio::Entry::new(&rel_path, Metadata::new(EntryMode::FILE))
-        } else if file_type.is_dir() {
+        let default_meta = self.op.metakey() == Metakey::Mode;
+
+        let (metadata, is_dir) = if default_meta {
+            let ft = de.file_type().map_err(new_std_io_error)?;
+            get_metadata(self.op.metakey(), None, Some(ft))?
+        } else {
+            let fs_meta = de.metadata().map_err(new_std_io_error)?;
+            get_metadata(self.op.metakey(), Some(fs_meta), None)?
+        };
+
+        let p = if is_dir {
             // Make sure we are returning the correct path.
-            oio::Entry::new(&format!("{rel_path}/"), Metadata::new(EntryMode::DIR))
+            &format!("{rel_path}/")
         } else {
-            oio::Entry::new(&rel_path, Metadata::new(EntryMode::Unknown))
+            &rel_path
         };
 
-        Ok(Some(entry))
+        Ok(Some(oio::Entry::new(p, metadata)))
     }
 }
+
+fn get_metadata(
```

Comment:
Hi, I feel like this helper function is a bit overdesigned. How about writing the logic directly? OpenDAL doesn't have a code style yet, but we tend to optimize for readability and avoid unnecessary abstraction, which can help a lot.

Reply:
I meant to share the same logic in the `List` and `BlockingList`

Reply:
> I meant to share the same logic in the `List` and `BlockingList`

Yes, I understand your motivation. However, sharing code that is used only twice is not encouraged in OpenDAL.

Reply:
Get it, will follow your suggestion, thanks